src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs

```diff
@@ -478,47 +482,64 @@ private void StartTelemetryIfFunctionInvocation(IDictionary<string, object> stat
 
-                    if (activities != null && activities.Length == 1)
-                    {
-                        operation = _telemetryClient.StartOperation<RequestTelemetry>(activities[0]);
-                        operation.Telemetry.Name = functionName;
-                    }
-                    else
+                    if (activities != null)
                     {
-                        RequestTelemetry request = new RequestTelemetry
+                        if (activities.Length == 1)
                         {
-                            Name = functionName
-                        };
-
-                        operation = _telemetryClient.StartOperation(request);
-
-                        // if there is more than one link (batch dispatch in EventHub trigger)
-                        // we'll populate link information on the request telemetry, but don't touch parent for this request
-                        PopulateLinks(activities, request);
+                            operation = _telemetryClient.StartOperation<RequestTelemetry>(activities[0]);
+                            operation.Telemetry.Name = functionName;
+                        }
+                        else
+                        {
+                            operation = CreateRequestFromLinks(activities, functionName);
+                        }
 
-                        // If any of the links is sampled in (on upstream) we also preliminary sample in 
-                        // request telemetry and everything that happens in this request scope. 
-                        // There will be additional level of sampling applied to limit rate to one 
-                        // configured on this Function/WebJob
-                        if (request.ProactiveSamplingDecision == SamplingDecision.SampledIn)
+                        if (this.TryGetAverageTimeInQueueForBatch(activities, operation.Telemetry.Timestamp, out long enqueuedTime))
                         {
-                            Activity.Current.ActivityTraceFlags |= ActivityTraceFlags.Recorded;
+                            operation.Telemetry.Metrics["timeSinceEnqueued"] = enqueuedTime;
                         }
                     }
+                    else
+                    {
+                        operation = _telemetryClient.StartOperation<RequestTelemetry>(functionName);
+                    }
+
+                    var triggerDetails = stateValues.GetValueOrDefault<IDictionary<string, string>>(ScopeKeys.TriggerDetails);
+                    if (triggerDetails != null &&
+                        triggerDetails.TryGetValue("EventHubEndpoint", out var endpoint) &&
+                        triggerDetails.TryGetValue("EventHubName", out var hubName))
+                    {
+                        operation.Telemetry.Source = endpoint.EndsWith("/") ? string.Concat(endpoint, hubName) : string.Concat(endpoint, "/", hubName);
+                    }
```

Comment:
I don't like the logger having special knowledge about EventHubs properties here. Can these be added to the Activity's tags in the EH extension and then they will just get pushed to the RequestTelemetry automatically?

Reply:
Oh... I see you want this to go with the `operation.Telemetry.Source` rather than with the custom properties. Could we have some "well-known" property in `LogConstants` like "Source" that we look up special for this, and have EH set that instead?

Reply:
Source is a special field for AppInsights UX, this is the place that is intended to have information about caller. 
Basically using `Source` is critical part for the UX. EventGrid or ServiceBus or Storage Queues should use `Source` in the same way as  EventHub does

Custom dimensions are a property bag that, in general, UX do not use - they are more expensive to access and group by.

Reply:
we can possibly standardize trigger details used to populate info about queue (not specific to the EventHub)

- `endpoint` will contain URI of the EventHub/ServiceBus/storage account/etc
- `entity` will contain entity name within account (hub/bus/queue/topic/container name).

Reply:
I think I'd feel better about that... having EH-specific stuff in here for logging feels wrong. Ideally the extension would handle this, but I can't think of any way for the EH extension to log this itself unless it registered its own `ITelemetryProcessor`.

Reply:
I've done the trigger details attributes not-specific to EH. Registering `ITelemetryProcessor` would mean taking dependency on AI packages and I'd prefer to avoid it.